ddl/multi_schema_change.go

```diff
@@ -56,18 +59,191 @@ func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
-	err := checkOperateSameColAndIdx(info)
-	if err != nil {
-		return err
+func onMultiSchemaChange(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	if job.MultiSchemaInfo.Revertible {
+		// Handle the rolling back job.
+		if job.IsRollingback() {
+			// Rollback/cancel the sub-jobs in reverse order.
+			for i := len(job.MultiSchemaInfo.SubJobs) - 1; i >= 0; i-- {
+				sub := job.MultiSchemaInfo.SubJobs[i]
+				if isFinished(sub) {
+					continue
+				}
+				proxyJob := cloneFromSubJob(job, sub)
+				ver, err = w.runDDLJob(d, t, proxyJob)
+				mergeBackToSubJob(proxyJob, sub)
+				if i == 0 && isFinished(sub) {
+					job.State = model.JobStateRollbackDone
+				}
+				return ver, err
+			}
+			// The last rollback/cancelling sub-job is done.
+			job.State = model.JobStateRollbackDone
+			return ver, nil
+		}
+
+		// The sub-jobs are normally running.
+		// Run the first executable sub-job.
+		for _, sub := range job.MultiSchemaInfo.SubJobs {
+			if !sub.Revertible || isFinished(sub) {
+				// Skip the sub-jobs which related schema states
+				// are in the last revertible point.
+				// If a sub job is finished here, it should be a noop job.
+				continue
+			}
+			proxyJob := cloneFromSubJob(job, sub)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			mergeBackToSubJob(proxyJob, sub)
+			handleRevertibleException(job, sub, proxyJob.Error)
+			return ver, err
+		}
+
+		// Save tblInfo and subJobs for rollback
+		tblInfo, _ := t.GetTable(job.SchemaID, job.TableID)
+		subJobs := make([]model.SubJob, len(job.MultiSchemaInfo.SubJobs))
+		// Step the sub-jobs to the non-revertible states all at once.
+		for i, sub := range job.MultiSchemaInfo.SubJobs {
+			if isFinished(sub) {
+				continue
+			}
+			subJobs[i] = *sub
+			proxyJob := cloneFromSubJob(job, sub)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			mergeBackToSubJob(proxyJob, sub)
+			if err != nil || proxyJob.Error != nil {
+				for j := i - 1; j >= 0; j-- {
+					job.MultiSchemaInfo.SubJobs[j] = &subJobs[j]
+				}
+				handleRevertibleException(job, sub, proxyJob.Error)
+				return updateVersionAndTableInfo(d, t, job, tblInfo, true)
+			}
+		}
+		// All the sub-jobs are non-revertible.
+		job.MultiSchemaInfo.Revertible = false
+		return ver, err
+	}
+	// Run the rest non-revertible sub-jobs one by one.
+	for _, sub := range job.MultiSchemaInfo.SubJobs {
+		if isFinished(sub) {
+			continue
+		}
+		proxyJob := cloneFromSubJob(job, sub)
+		ver, err = w.runDDLJob(d, t, proxyJob)
+		mergeBackToSubJob(proxyJob, sub)
+		return ver, err
 	}
+	job.State = model.JobStateDone
+	return ver, err
+}
 
-	err = checkVisibleColumnCnt(t, len(info.AddColumns), len(info.DropColumns))
+func isFinished(job *model.SubJob) bool {
+	return job.State == model.JobStateDone ||
+		job.State == model.JobStateRollbackDone ||
+		job.State == model.JobStateCancelled
+}
+
+func cloneFromSubJob(job *model.Job, sub *model.SubJob) *model.Job {
+	return &model.Job{
```

Comment:
It's likely to miss some fields in the feature, how to avoid it?

Reply:
I've add some comments to the definition of `model.Job`:

```
// NOTE: To add a field, please make sure that the following methods work as expected:
// - ddl.cloneFromSubJob
```